README.md
# simplesamlphp-module-aup
SimpleSAMLphp module for handling AUP information

## Configuration

The following configuration options are available for the UpdateAUP filter which allows users to update agreement to relevant AUPs:
  - `aupApiEndpoint`: Required, the API endpoint for storing AUP agreements
  - `aupListEndpoint`: Required, the endpoint for AUPs list page of the user
  - `apiUsername`: Required, the username of the API user
  - `apiPassword`: Required, the password of the API user
  - `apiTimeout` : Optional, API timeout value in seconds (default: 15)
  - `userIdAttribute` : Optional, a string containing the name of the attribute whose value will be used to check if it is in userIdBlacklist.
  - `spBlacklist`: Optional, an array of strings that contains the SPs that the module will skip the process.
  - `userIdBlacklist`: Optional, an array of strings that contains the userIds for which the module will skip the AUP process. In order to be activated a value is required to userIdAttribute option.

### Example configuration

```
     'authproc' => [
        ...
        '82' => [
             'class' => 'aup:UpdateAUP',
             'aupApiEndpoint' => '',
             'aupListEndpoint' => '',
             'apiUsername' => '',
             'apiPassword' => '',
             'apiTimeout'  => 15,
             'userIdAttribute' => '',
             'spBlacklist' => [],
             'userIdBlacklist' => []
        ],
```

###  View Loader

The filter could exploit a pure css loader/spinner specified in the `loader` configuration option of the global theme used by the SimpleSAMLphp installation; if no `loader` configuration option is found (e.g. when using the default SimpleSAMLphp theme) then no loader will be shown.
The loader element file must be included under `/includes` directory.
e.g.
```php
$this->includeAtTemplateBase('includes/' . $loader . '.php');
```

## Compatibility matrix

This table matches the module version with the supported SimpleSAMLphp version.

| Module |  SimpleSAMLphp |
|:------:|:--------------:|
| v1.x   | v1.14          |
| v2.x   | v1.17+         |

# License

Licensed under the Apache 2.0 license, for details see `LICENSE`.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v2.0.3] - 2021-01-25

### Fixed

- Namespace errors

## [v2.0.2] - 2021-01-22

### Added

- Option for configuring the timeout for AUP update API calls

### Changed

- Load loader load from global SimpleSAMLphp theme
- Decouple css and javascript
- Refactor template

## [v1.0.3] - 2021-01-21

### Added

- Option for configuring the timeout for AUP update API calls

### Changed

- Load loader load from global SimpleSAMLphp theme
- Decouple css and javascript
- Refactor template

## [v2.0.1] - 2021-01-17

### Fixed

- Include AUPs never agreed before to the list of AUPs that need to be (re)accepted

## [v2.0.0] - 2021-01-14

This version is compatible with [SimpleSAMLphp v1.17](https://simplesamlphp.org/docs/1.17/simplesamlphp-changelog)

### Changed

- Compatibility with [SimpleSAMLphp v1.17](https://simplesamlphp.org/docs/1.17/simplesamlphp-changelog)

## [v1.0.2] - 2021-01-15

### Fixed

- Include AUPs never agreed before to the list of AUPs that need to be (re)accepted

## [v1.0.1] - 2021-01-14

### Changed

- Improve handling of missing AUP information
- Improve logging

## [v1.0.0] - 2021-01-12

This version is compatible with [SimpleSAMLphp v1.14](https://simplesamlphp.org/docs/1.14/simplesamlphp-changelog)

### Added

- Initial version of UpdateAUP authproc filter for allowing users to update agreement to relevant AUPs


www/aup_in_form.php

<?php
/**
 * Review Updated AUP Form
 *
 * This script displays a page to the user, which requests that the user
 * should review the updated AUPs.
 *
 * @package SimpleSAMLphp
 */
/**
 * Explicit instruct attribute selection page to send no-cache header to browsers to make
 * sure the users attribute information are not store on client disk.
 *
 * In an vanilla apache-php installation is the php variables set to:
 *
 * session.cache_limiter = nocache
 *
 * so this is just to make sure.
 */
session_cache_limiter('nocache');
$globalConfig = SimpleSAML\Configuration::getInstance();
if (!array_key_exists('StateId', $_REQUEST)) {
    throw new SimpleSAML\Error\BadRequest(
      'Missing required StateId query parameter.'
    );
}
$id = $_REQUEST['StateId'];
/* Restore state */
$state = SimpleSAML\Auth\State::loadState($id, 'aup_state');

// The user has pressed the yes-button
// The resumeProcessing function needs a ReturnUrl or a ReturnCall in order to proceed
if (array_key_exists('yes', $_REQUEST)) {
    SimpleSAML\Logger::debug("[aup] REQUEST". var_export($_REQUEST, true));
    $url = $state['aup:aupApiEndpoint'];
    foreach($state['aup:changedAups'] as $aup) {
        SimpleSAML\Logger::debug("[aup] Changed AUPS:". $aup['id']);
        SimpleSAML\Logger::debug("[aup] User Id:".   $state["rciamAttributes"]["registryUserId"]);

        if(!empty($_REQUEST['terms_and_conditions_'.$aup['id']])){
            // Make the post requests
            addCoTAndCAgreement($state["rciamAttributes"]["registryUserId"], $aup['id'], $url, $state['aup:apiUsername'], $state['aup:apiPassword'], $state['aup:apiTimeout']);
        }
    }
    if (array_key_exists('aup:changedAups', $state)) {
        unset($state['aup:changedAups']);
    }
    if (array_key_exists('aup:aupListEndpoint', $state)) {
        unset($state['aup:aupListEndpoint']);
    }
    if (array_key_exists('aup:aupApiEndpoint', $state)) {
        unset($state['aup:aupApiEndpoint']);
    }
    if (array_key_exists('aup:apiUsername', $state)) {
        unset($state['aup:aupUsername']);
    }
    if (array_key_exists('aup:apiPassword', $state)) {
        unset($state['aup:aupPassword']);
    }
    SimpleSAML\Auth\ProcessingChain::resumeProcessing($state);
    if (array_key_exists('aup:apiTimeout', $state)) {
            unset($state['aup:apiTimeout']);
    }
}

// Make, populate and layout informed failure consent form
$t = new SimpleSAML\XHTML\Template($globalConfig, 'aup:aup_in_form.tpl.php');
$t->data['yesTarget'] = SimpleSAML\Module::getModuleURL('aup/aup_in_form.php');
$t->data['yesData'] = array('StateId' => $id);
$t->data['changedAups'] = $state['aup:changedAups'];
$t->data['aupListEndpoint'] = $state['aup:aupListEndpoint'];
$t->show();

function addCoTAndCAgreement($coPersonId, $coTAndCId, $url, $apiUser, $apiPass, $apiTimeout)
{
    // Construct my data
    $reqDataArr = array();
    $reqDataArr['RequestType'] = 'CoTAndCAgreements';
    $reqDataArr['Version'] = '1.0';
    $reqDataArr['CoTAndCAgreements'][0]['Version'] = '1.0';
    $reqDataArr['CoTAndCAgreements'][0]['CoTermsAndConditionsId'] = (string) ($coTAndCId);
    $reqDataArr['CoTAndCAgreements'][0]['Person']['Type'] = 'CO';
    $reqDataArr['CoTAndCAgreements'][0]['Person']['Id'] = $coPersonId;
    $req = json_encode($reqDataArr);
    $res = http('POST', $url, $req, $apiUser, $apiPass, $apiTimeout);
    return $res;
}

function http($method, $url, $data = null, $apiUser, $apiPass, $apiTimeout)
{

    $ch = curl_init($url);
    curl_setopt_array(
        $ch,
        array(
          CURLOPT_CUSTOMREQUEST => $method,
          CURLOPT_CONNECTTIMEOUT => $apiTimeout,
          CURLOPT_RETURNTRANSFER => true,
          CURLOPT_USERPWD => $apiUser . ":" . $apiPass,
        )
    );
    if (($method == "POST" || $method == "PUT") && !empty($data)) {
        curl_setopt($ch, CURLOPT_POSTFIELDS, $data);
        curl_setopt($ch, CURLOPT_HTTPHEADER, array(
                         'Content-Type: application/json',
                         'Content-Length: ' . strlen($data))
        );
    }

    // Send the request
    $response = curl_exec($ch);
    $http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);

    // Check for error
    if ($http_code !== 200 && $http_code !== 201 && $http_code !== 204 && $http_code !== 302 && $http_code !== 404) {
        SimpleSAML\Logger::error("[aup] save acceptance of aup failed. http: method=" // TODO error logging
        . var_export($method, true) . ", url=" . var_export($url, true)
        . ", data=" . var_export($data, true)
        . ": API call failed: HTTP response code: "
        . var_export($http_code, true) . ", error message: '"
        . var_export(curl_error($ch), true) . "'\n");
    }
    // Close session
    curl_close($ch);
    $result = json_decode($response);
    SimpleSAML\Logger::debug("[aup] api call for renew AUP http: result="
        . var_export($result, true));
    assert('json_last_error()===JSON_ERROR_NONE');
    return $result;
}

lib/Auth/Process/Client.php
<?php

/**
 * Authproc filter for informing the user for updated AUP(s) and asking them to re-accept them
 * in order to proceed
 *
 * Example configuration:
 *
 *    'authproc' => array(
 *       ...
 *       '82' => array(
 *            'class' => 'aup:Client',
 *            'aupApiEndpoint' => '',
 *            'aupListEndpoint' => '',
 *            'apiUsername' => '',
 *            'apiPassword' => '',
 *            'userIdAttribute' => '',
 *            'spBlacklist' => array(),
 *            'userIdBlacklist' => array()
 *       ),
 *
 * @author Nick Mastoris <nmastoris@admin.grnet.gr>
 */
class sspmod_aup_Auth_Process_Client extends SimpleSAML_Auth_ProcessingFilter
{

    public function __construct($config, $reserved)
    {
        parent::__construct($config, $reserved);
        $params = array(
            'aupApiEndpoint',
            'aupListEndpoint',
            'apiUsername',
            'apiPassword'
        );
        foreach ($params as $param) {
            if (!array_key_exists($param, $config)) {
                throw new SimpleSAML_Error_Exception(
                    'Missing required configuration parameter: ' .$param);
            }
            $this->config[$param] = $config[$param];
        }

        $optionals = array(
            'userIdAttribute',
            'spBlacklist',
            'userIdBlacklist'
        );
        foreach ($optionals as $optional) {
            if(!empty($config[$optional])) {
                $this->config[$optional] = $config[$optional];
            }
        }
    }

    /**
     * @param array $state
     */
    public function process(&$state)
    {
        if (!empty($this->config['spBlacklist']) && isset($state['SPMetadata']['entityid']) && in_array($state['SPMetadata']['entityid'], $this->config['spBlacklist'], true)) {
            SimpleSAML_Logger::debug("[aup] process: Skipping blacklisted SP ". var_export($state['SPMetadata']['entityid'], true));
            return;
        }
        // Check if user is in blacklist
        if(!empty($this->config['userIdAttribute']) && !empty($this->config['userIdBlacklist']) &&  !empty(array_intersect($state['Attributes'][$this->config['userIdAttribute']], $this->config['userIdBlacklist']))) {
            return;
        }
        try {
            SimpleSAML_Logger::info('[aup] process: ' . $this->config['userIdAttribute'] . ' ' . var_export($state['Attributes'][$this->config['userIdAttribute']],true));
            // Check if we have an updated aup
            $changed_aups = array();

            foreach ($state['rciamAttributes']['aup'] as $aup) {
                if (!empty($aup['agreed']) && $aup['version'] != $aup['agreed']['version']) {
                    $changed_aups[] = $aup;
                }
            }
            if (!empty($changed_aups) && !empty($state["rciamAttributes"]) && !empty($state["rciamAttributes"]["registryUserId"])) {
                    $state['aup:changedAups'] = $changed_aups;
                    $state['aup:aupListEndpoint'] = str_replace("%registryUserId%", $state["rciamAttributes"]["registryUserId"], $this->config['aupListEndpoint']);
                    $state['aup:aupApiEndpoint'] = $this->config['aupApiEndpoint'];
                    $state['aup:apiUsername'] = $this->config['apiUsername'];
                    $state['aup:apiPassword'] = $this->config['apiPassword'];

                    $id = SimpleSAML_Auth_State::saveState($state, 'aup_state');
                    $url = SimpleSAML_Module::getModuleURL('aup/aup_in_form.php');
                    \SimpleSAML\Utils\HTTP::redirectTrustedURL($url, array('StateId' => $id));
          }
          return;
        } catch (\Exception $e) {
            $this->showException($e);
        }

    }

    /**
      * @param $e
      * @throws Exception
      */
    private function showException($e)
    {
        $globalConfig = SimpleSAML_Configuration::getInstance();
        $t = new SimpleSAML_XHTML_Template($globalConfig, 'aup:exception.tpl.php');
        $t->data['e'] = $e->getMessage();
        $t->show();
        exit();
    }
}


templates/aup_in_form.tpl.php
<?php

/**
 * Template form for attribute selection.
 *
 * Parameters:
 * - 'yesTarget': Target URL for the yes-button. This URL will receive a POST request.
 * - 'yesData': Parameters which should be included in the yes-request.
 * - 'changedAups': AUPs that have been updated.
 * - 'aupListEndpoint': AUP List Endpoint showing all AUPs for user.
 *
 * @package SimpleSAMLphp
 */
assert('is_string($this->data["yesTarget"])');
assert('is_array($this->data["yesData"])');
assert('is_array($this->data["changedAups"])');

// Parse parameters
if (array_key_exists('changedAups', $this->data)) {
  $aups = $this->data['changedAups'];
}
if (array_key_exists('aupListEndpoint', $this->data)) {
  $aupListEndpoint = $this->data['aupListEndpoint'];
}

// Get Configuration and set the loader
$globalConfig = SimpleSAML_Configuration::getInstance();
$theme_use = $globalConfig->getString('theme.use', 'default');
if ($theme_use !== 'default') {
    $theme_config_file = 'module_' . explode(':', $theme_use)[0] . '.php';
    $themeConfig       = SimpleSAML_Configuration::getConfig($theme_config_file);
    $loader = $themeConfig->getValue('loader');
    if (!empty($loader)) {
        $this->includeAtTemplateBase('includes/' . $loader . '.php');
    }
}

$this->data['jquery'] = array('core' => true, 'ui' => true, 'css' => true);
// Configure HTML head
$this->data['head'] = header("Expires: Thu, 19 Nov 1981 08:52:00 GMT");                //Date in the past in order not to keep cache if someone goes back to that page
$this->data['head'] .= header("Cache-Control: no-store, no-cache, must-revalidate");   //HTTP/1.1
// Include custom javascript
$this->data['head'] .= '<script type="text/javascript" src="/proxy/module.php/aup/resources/js/aup.js"></script>';
// Include custom style
$this->data['head'] .= '<link rel="stylesheet" type="text/css" href="resources/css/aup.css" />' . "\n";


$this->includeAtTemplateBase('includes/header.php');
?>

    <h1 class="text-center" style="padding-bottom: 0.8em; font-size: xx-large; font-weight: 700;">
        <?php print $this->t('{aup:aup:updated_aup_notice}'); ?>
    </h1>
    <h2 class="text-center">
        <?php print $this->t('{aup:aup:updated_aup_title}'); ?>
    </h2>
    <div class="text-center" style="font-size:1.2em; margin-top:20px; line-height: 1.6em;">
        <?php print $this->t('{aup:aup:updated_aup_description}'); ?>
    </div>

    <!--  Form that will be sumbitted on Yes -->
    <form style="display: inline; margin: 0px; padding: 0px" action="<?php print htmlspecialchars($this->data['yesTarget']); ?>">

        <div style="font-size: 1em;margin-top:2em;">
              <?php foreach ($aups as $aup): ?>
                  <div class="row aup_rows" style="padding:7px 0px">
                      <div class="col-lg-5 col-lg-offset-2 col-md-offset-2 col-md-5 col-sm-7">
                          <h3>
                              <a class="aup_content"
                                 data-description="<?php print $aup['description']; ?>"
                                 data-url="<?php print $aup['url']; ?>">
                                <?php print $aup['description']; ?>
                              </a>
                          </h3>
                      </div>
                      <div class="col-lg-5 col-md-5 col-sm-5">
                          <h3>
                              <input type="checkbox" name="terms_and_conditions_<?php print $aup['id'] ?>"/>
                              <span style="font-size: 0.9em"> I Agree</span>
                          </h3>
                      </div>
                  </div>
              <?php endforeach; ?>
        </div>
        <p style="margin:1em 1em 5em " class="text-center">
            <?php foreach ($this->data['yesData'] as $name => $value): ?>
            <input type="hidden"
                   name="<?php print htmlspecialchars($name); ?>"
                   value="<?php print htmlspecialchars($value); ?>" />
            <?php endforeach; ?>
            <button type="submit" name="yes" disabled="disabled"
                    class="ssp-btn btn ssp-btn__action ssp-btns-container--btn__left text-uppercase" id="yesbutton">
              <?php print htmlspecialchars($this->t('{aup:aup:yes}')) ?>
            </button>
        </p>
    </form>
    <p class="text-center" style="margin-top:20px;margin-bottom:50px">
        <?php print $this->t(
        '{aup:aup:updated_aup_more_information}',
        array('%HERE%' => '<a href="'.$aupListEndpoint.'">here</a>')
      ) ?>
    </p>
<?php
// Include Modal
$this->includeAtTemplateBase('includes/modal.php');
// Include Footer
$this->includeAtTemplateBase('includes/footer.php');


lib/Auth/Process/UpdateAUP.php
<?php

/**
 * Authproc filter for informing the user for updated AUP(s) and asking them to re-accept them
 * in order to proceed
 *
 * Example configuration:
 *
 *    'authproc' => [
 *       ...
 *       '82' => [
 *            'class' => 'aup:UpdateAUP',
 *            'aupApiEndpoint' => '',
 *            'aupListEndpoint' => '',
 *            'apiUsername' => '',
 *            'apiPassword' => '',
 *            'apiTimeout' => 15,
 *            'userIdAttribute' => '',
 *            'spBlacklist' => [],
 *            'userIdBlacklist' => []
 *       ],
 *
 * @author Nick Mastoris <nmastoris@admin.grnet.gr>
 */
namespace SimpleSAML\Module\aup\Auth\Process;

use SimpleSAML\Auth\State;
use SimpleSAML\Logger;
use SimpleSAML\Module;
use SimpleSAML\Utils\HTTP;
use SimpleSAML\Error;

class UpdateAUP extends \SimpleSAML\Auth\ProcessingFilter
{

    public function __construct($config, $reserved)
    {
        // Default value for api timeout
        $this->config['apiTimeout'] = 15;

        parent::__construct($config, $reserved);
        $params = array(
            'aupApiEndpoint',
            'aupListEndpoint',
            'apiUsername',
            'apiPassword'
        );
        foreach ($params as $param) {
            if (!array_key_exists($param, $config)) {
                throw new Error\Exception(
                    'Missing required configuration parameter: ' .$param);
            }
            $this->config[$param] = $config[$param];
        }

        $optionals = array(
            'userIdAttribute',
            'spBlacklist',
            'userIdBlacklist',
            'apiTimeout'
        );
        foreach ($optionals as $optional) {
            if(!empty($config[$optional])) {
                $this->config[$optional] = $config[$optional];
            }
        }
    }

    /**
     * @param array $state
     */
    public function process(&$state)
    {
        if (!empty($this->config['spBlacklist']) && isset($state['SPMetadata']['entityid']) && in_array($state['SPMetadata']['entityid'], $this->config['spBlacklist'], true)) {
            Logger::debug("[aup:UpdateAUP] Skipping blacklisted SP ". var_export($state['SPMetadata']['entityid'], true));
            return;
        }
        // Check if user is in blacklist
        if ((!empty($this->config['userIdAttribute']) && !empty($state['Attributes'][$this->config['userIdAttribute']]) && !empty($this->config['userIdBlacklist']) && !empty(array_intersect(
            $state['Attributes'][$this->config['userIdAttribute']],
            $this->config['userIdBlacklist'])))) {
            Logger::debug("[aup:UpdateAUP] Skipping blacklisted user with id " . var_export($state['Attributes'][$this->config['userIdAttribute']], true));
            return;
        }
        // Check if $state['rciamAttributes']['aup'] is empty
        if (empty($state['rciamAttributes']['aup'])) {
            Logger::debug("[aup:UpdateAUP] No AUP information found in state - skipping");
            return;
        }
        try {
            // Check if there are updated aup(s)
            $changed_aups = array();

            foreach ($state['rciamAttributes']['aup'] as $aup) {
                if (empty($aup['agreed']) || $aup['version'] != $aup['agreed']['version']) {
                    $changed_aups[] = $aup;
                }
            }
            if (!empty($changed_aups) && !empty($state["rciamAttributes"]) && !empty($state["rciamAttributes"]["registryUserId"])) {
                    $state['aup:changedAups'] = $changed_aups;
                    $state['aup:aupListEndpoint'] = str_replace("%registryUserId%", $state["rciamAttributes"]["registryUserId"], $this->config['aupListEndpoint']);
                    $state['aup:aupApiEndpoint'] = $this->config['aupApiEndpoint'];
                    $state['aup:apiUsername'] = $this->config['apiUsername'];
                    $state['aup:apiPassword'] = $this->config['apiPassword'];
                    $state['aup:apiTimeout'] = $this->config['apiTimeout'];

                    $id = State::saveState($state, 'aup_state');
                    $url = Module::getModuleURL('aup/aup_in_form.php');
                    HTTP::redirectTrustedURL($url, array('StateId' => $id));
          }
          return;
        } catch (\Exception $e) {
            $this->showException($e);
        }

    }

    /**
      * @param $e
      * @throws Exception
      */
    private function showException($e)
    {
        $globalConfig = SimpleSAML\Configuration::getInstance();
        $t = new SimpleSAML\XHTML\Template($globalConfig, 'aup:exception.tpl.php');
        $t->data['e'] = $e->getMessage();
        $t->show();
        exit();
    }
}
